### PR TITLE
Aggregators in the IR (now with Map, Filter, & FlatMap)

### PR DESCRIPTION
**Pull request by danking**
cc: @catoverdrive @tpoterba 

I collapsed two PRs into one and made major changes, so this needs a fresh look.

Some high level direction:

 - RegionValueAggregator is an interface that doesn't take `Any` as the seqOp argument. All of  them have methods that can be called directly with the right input type. I emit code to call such methods.

 - TransformedRegionValueAggregator is just a little shim class that lets me override the seqOp of another aggregator with an AsmFunction4 (the result of compiling)

 - ZippedRegionValueAggregator just puts a bunch of RegionValueAggregators together (just foreach on the child aggregators) and puts all their results into an RV Struct. This RV Struct is referenced in the non-aggregating IR code (I've been calling the fragment that remains after aggs are removed "stage0" and calling aggregation "stage1"). Passing all the aggregations directly as arguments to stage0 seems hard/impossible because the number of aggregations isn't known at scala-compile-time, but we stitch together the aggregations and the stage0 compiled IR in scala.

 - I added 4 IRs: AggIn (this is `gs`),  AggMap, AggFilter, AggFlatMap, AggSum. I don't quite know how to add more aggregators. I think I will create new IRs for each one. Or maybe I can get a better interface. Comments welcome.

 - `compileAgg2` is a CPS-style compiler that lays out straight-line code that corresponds to aggregation

 - The interface to generated aggregations isn't very nice right now. The user passes in a struct that contains the element and the scope values. When running an aggregator, the element is extracted first and then passed around (without the rest of the scope). References to the scope become `GetField` IRs that find the appropriate field in the argument. Using this is a bit hacky because we should mutate the fields of the scope struct as we loop over `gs` (e.g. `sa`). However, this is essentially what we do now (EvalContext is basically the scope struct), so I'm not so bothered by it. My tests are kind of hard to follow as a result.

### CODE REVIEW

File: src/main/scala/is/hail/expr/ir/ExtractAggregators.scala [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+          case _: TBoolean => ((v: Code[_], mv: Code[Boolean]) =>
+            mv.mux(
+              Code.checkcast[RegionValueSumBooleanAggregator](aggregator).invoke[Boolean, Boolean, Unit](
+                "seqOp", coerce[Boolean](defaultValue(tElement)), true),
+              Code.checkcast[RegionValueSumBooleanAggregator](aggregator).invoke[Boolean, Boolean, Unit](
+                "seqOp", coerce[Boolean](v), false)),
+            new RegionValueSumBooleanAggregator())
+          case _: TInt32 => ((v: Code[_], mv: Code[Boolean]) =>
+            mv.mux(
+              Code.checkcast[RegionValueSumIntAggregator](aggregator).invoke[Int, Boolean, Unit](
```
Comment (danking) [inline src/main/scala/is/hail/expr/ir/ExtractAggregators.scala]: Kind of unhappy about all this duplication. I think to avoid it I have to work one level deeper in Code by explicitly passing `TypeInfo` to the `Invokeable` stuff.

Comment (cseed) [reply src/main/scala/is/hail/expr/ir/ExtractAggregators.scala]: It is unnecessary.  You can just cast to `RegionValueAggregator` in all cases since  you're calling a function from its interface.  Just the constructor sum aggregator object needs to vary.

**Comment by cseed** [comment]
This is better, but I want to go more extreme:

 - get rid of TransformedRegionValueAggregator and ZippedRegionValueAggregator.  This is a compiler backend.  Too much abstraction in your output!  Let's compile that shit away.
 - ExtractAggregators should return an Array[AggSum].  These are expressions ending in AggSum containing aggregator operations (filter, map, etc.) defined in terms of the aggregated element and associated context.
 - Add a function AggSum => RegionValueAggregator.  This is the way to generalize: make AggSum into Agg(op: AggOp) where AggOp (like unary and binary op) are all the possible aggregator types, and there is a function that maps the op to the corresponding RegionValueAggregator*
 - compiling the Array[AggSum] should product a function that takes the array of aggregators and a single value (with context) of the collection we're aggregating over and updates them with that element.

*I think you need an array of arguments to handle things like call_stats which are evaluated in the aggregator scope (the only scope available to evaluate something).

Imagine you have `gs.filter(g => g.GT.isHet()).map(g => g.DP).sum() + gs.flatMap(g => g.PL).sum()`.

The Array[AggSum] will be

```
  Array(AggSum(AggMap(
    AggFilter(AggIn(...), 
      "g", g.GT.isHet()),
      "g", (getField (Ref "g") "DP"),
    AggSum(AggFlatMap(AggIn(...),
      "g", (getField (Ref "g") "PL)))
```

The generated function should look like:

```
def f(aggs: Array[AggSum], region: MemoryBuffer, g: Long, mg: Boolean, ...) {
  if (g.GT.isHet()) { // RV-ified, of course
    val DP = g.DP // actually fieldOffset
    aggs(0).seqOp(DP)
  }
  for (PLi in g.PL) { // actually elementOffset
    aggs(1).seqOp(PLi)
  }
}
```

This is straight-line and should be fast.  It immediately allows you to do common subexpression elimination on aggregator prefixes which is something that is quite common, that is, if you have `gs.filter(g => g.isHet).map(g => g.DP).mean() < 10 gs.filter.map(g => g.GQ).mean() < 50` then in the aggregation function you can do:

```
  if (g.isHet) { // only test once!
    aggs(0).seqOp(g.DP)
    aggs(1).seqOp(g.GQ)
  }
```

File: src/main/scala/is/hail/expr/ir/ExtractAggregators.scala (hunk 1 above)
Comment (danking) [reply src/main/scala/is/hail/expr/ir/ExtractAggregators.scala]: I must misunderstand something because I'm pretty sure I'm _not_ calling a function from `RegionValueAggregator`'s interface. For example, here I call the `RegionValueSumIntAggregator`'s `seqOp(int, boolean)` method. The `RegionValueAggregator` interface takes an offset to a value. I didn't want to unnecessarily load an intermediate value into the region just so I can get an offset to hand to the aggregator.

Comment (cseed) [reply src/main/scala/is/hail/expr/ir/ExtractAggregators.scala]: Sorry, I thought you were calling the `seqOp(MemoryBuffer, Long, Boolean)` variant.  Is there any reason you can't do that?

Also, the mux here seems unnecessary.  Isn't `v` some dummy value already when `mv` is false?

Comment (danking) [reply src/main/scala/is/hail/expr/ir/ExtractAggregators.scala]: I can, but an intermediate value (any non-compound thing) will have to be written into the region so as to get an offset that I can pass to that interface. It seems silly to, for example, load the result of simple arithmetic (performed on the JVM stack) into the region and then unload it on the other side of the RegionValueAggregator interface.

Executing `v` when `mv` is `true` seems like a bad idea. What if `v` is `ArrayLength(NA(TArray(TInt32())))`. If the region is empty, then I'll try to loadLength from an array at offset 0, but `loadInt` requires the region is at least 4 bytes large. I think the crux of the issue is that we don't get undefined behavior when we read bad memory addresses, we get assertion errors. An even more problematic example is `ArrayRef(NA(TArray(TInt32())), I32(0)`. The offset to the missing array will get dummy value 0, and `loadElement` will happily try to read the bytes at 0 as the length of the array which of course could lead to read attempts outside the region.

Comment (cseed) [reply src/main/scala/is/hail/expr/ir/ExtractAggregators.scala]: You're right, my bad.  I was thinking the thing had an offset already and you should hold on to it, but that's clearly wrong because it might be the intermediate result of an expression.  This seems like the standard type specialization duplication.  Maybe you could do some type-driven seqOp lookup to avoid the duplication, but honestly, this looks fine to me now.

This comment on Compile makes the second point clear:

 > gaurd the the evaluation of value by missingness

It does make me wonder if this is the right choice.  In general, you had an if in the precompute Code to test the missingness of the inputs, and now you force yourself to have a second one in the consumption.  I'm OK with coming back to this when we work on optimized the generated code.

Comment (danking) [reply src/main/scala/is/hail/expr/ir/ExtractAggregators.scala]: The problem arises with values that _must_ do some computation to know if they are missing. For example, an array reference must know the value of the array to know if it is itself missing. I can either:

 - duplicate the evaluation of (and the bytecode for) the array

 - duplicate the missingness check

---

I think this situation would be resolved if the compiler was "CPS'ed" (using two "continuations"). The ArrayRef implementation would lay down the byte code to evaluate the array, then mux on the element's missingness:
 -  in the missing case it calls the "My value is missing" "continuation" (whose type is `() => Code[_]`)
 - if the element was non-missing, call the "here's my value" continuation (whose type is `(Code[_]) => Code[_]`) with the byte code to load the element

This mirrors how one would handle exceptions in a simple CPS'ed Scheme compiler, which is kinda how missingness works anyway.

Comment (cseed) [reply src/main/scala/is/hail/expr/ir/ExtractAggregators.scala]: I agree with your second comment that compiling with dual continuations is likely to generate better code.

I don't fully understand the first case.  You can just have a default initialized value for the array element value.  For example, imagine you're compiling `gs.map(g => a[g.DP]).sum()`, then you can do:

```
  var e, me // the element of a, default-initialized
  // evaluate a, ma
  if (ma)
    me = false
  else
    e = a(a.DP)
    me = true
  aggs(0).seqOp(e, me)  // don't need a second test because e is ignored if me = true
```

Comment (cseed) [reply src/main/scala/is/hail/expr/ir/ExtractAggregators.scala]: This (two continuation compilation) is essentially what I tried to do for Booleans in `Code`.  It's much better because you don't actually need branches on the intermediate missing bits and will almost certainly generate better code.  It dominates the usual approach because you can always set a missing bit on the two branches and jump to common code (like if you need to pass a value out or store it.)

Comment (danking) [reply src/main/scala/is/hail/expr/ir/ExtractAggregators.scala]: The byte code example you give works if I always pass values via local variables. In such a situation I can always set the value to a default value in the setup code that runs first.

One of the design goals I followed was that the compiler only stores a value in a local variable if it a) is the bound value in a let or b) is _definitely_ used more than once. For example, I don't store the value of `ArrayRef` in a local variable, (it's just a `region.loadType(tarray.loadElement(...))`), so I'd have to duplicate the missingness check in the code for the value to return a default value.

Storing all returned values in local variables seems a little silly, especially in large compositions of arithmetic expressions where every binary operation will get stored and then loaded.

The only place this is an issue is at method call boundaries like this (which never occur in regular IR) because method arguments are evaluated eagerly.

File: src/test/scala/is/hail/expr/ir/ExtractAggregatorsSuite.scala [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+    rvb.start(t)
+    rvb.startStruct()
+    t.fields.map(_.typ).zip(a).foreach { case (t, agg) =>
+      rvb.addRegionValue(t, region, agg.result(region))
+    }
+    rvb.endStruct()
+    rvb.end()
+  }
+
+  private def loadIRIntermediate(typ: Type): (MemoryBuffer, Long) => Any = typ match {
```
Comment (danking) [inline src/test/scala/is/hail/expr/ir/ExtractAggregatorsSuite.scala]: I wasn't sure this was appropriate to put on a Region, it should only be used by tests.

File: src/main/scala/is/hail/annotations/aggregators/RegionValueSumAggregator.scala [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+import is.hail.expr.RegionValueAggregator
+
+import scala.reflect.ClassTag
+
+object RegionValueSumAggregator {
+  private def seqOp[Agg >: Null : ClassTag : TypeInfo, T : ClassTag](t: Type): (Code[RegionValueAggregator], Code[_], Code[Boolean]) => Code[Unit] =
+  { (rva: Code[RegionValueAggregator], v: Code[_], mv: Code[Boolean]) =>
+    mv.mux(
+      Code.checkcast[Agg](rva).invoke[T, Boolean, Unit]("seqOp", coerce[T](defaultValue(t)), true),
+      Code.checkcast[Agg](rva).invoke[T, Boolean, Unit]("seqOp", coerce[T](v), false)) }
```
Comment (danking) [inline src/main/scala/is/hail/annotations/aggregators/RegionValueSumAggregator.scala]: I thought this was a cute way to get what we wanted.

File: src/test/scala/is/hail/expr/ir/ExtractAggregatorsSuite.scala [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+  }
+
+  private def compileStage0[R: TypeInfo](ir: IR): AsmFunction3[Region, Long, Boolean, R] = {
+    val fb = FunctionBuilder.functionBuilder[Region, Long, Boolean, R]
+    // nb: inference is done by stage1
+    Compile(ir, fb)
+    fb.result()()
+  }
+
+  private def run[T : HailRep : TypeInfo, U : HailRep : TypeInfo, R : TypeInfo](region: Region, ir: IR, aOff: Long, scope1: Int => U): R = {
```
Comment (danking) [inline src/test/scala/is/hail/expr/ir/ExtractAggregatorsSuite.scala]: Runs a possibly aggregating `ir` using the array at `aOff` in `region` as the aggregable. There's one name in scope called "scope0" and its value is determined by `scope0: Int => U`.

File: src/main/scala/is/hail/expr/ir/TypeToTypeInfo.scala [hunk 5]
```diff
@@ -0,0 +1,15 @@
+package is.hail.expr.ir
+
+import is.hail.expr._
+import is.hail.asm4s._
+
+object TypeToTypeInfo {
+  def apply(t: Type): TypeInfo[_] = t match {
```
Comment (cseed) [inline src/main/scala/is/hail/expr/ir/TypeToTypeInfo.scala]: Is this fundamental?  I think so (assert) or match on t.fundamental

File: src/main/scala/is/hail/expr/ir/TypeToTypeInfo.scala [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+    case _: TInt32 => typeInfo[Int]
+    case _: TInt64 => typeInfo[Long]
+    case _: TFloat32 => typeInfo[Float]
+    case _: TFloat64 => typeInfo[Double]
+    case _: TBoolean => typeInfo[Boolean]
+    case _ => typeInfo[Long] // reference types
```
Comment (cseed) [inline src/main/scala/is/hail/expr/ir/TypeToTypeInfo.scala]: And enumerate struct, array.

File: src/main/scala/is/hail/expr/ir/package.scala [hunk 7]
```diff
@@ -3,5 +3,12 @@ package is.hail.expr
...
 
 package object ir {
-  def coerce[T](x: Code[_]): Code[T] = x.asInstanceOf[Code[T]]
+  def defaultValue(t: Type): Code[_] = t match {
+    case _: TBoolean => false
+    case _: TInt32 => 0
+    case _: TInt64 => 0L
+    case _: TFloat32 => 0.0f
+    case _: TFloat64 => 0.0
+    case _ => 0L // reference types
```
Comment (cseed) [inline src/main/scala/is/hail/expr/ir/package.scala]: Same concerns here.

File: src/main/scala/is/hail/asm4s/GenericTypeInfo.scala [hunk 8]
```diff
@@ -58,6 +60,8 @@ final case class GenericTypeInfo[T : TypeInfo]() extends MaybeGenericTypeInfo[T]
       Code.boxBoolean(x.asInstanceOf[Code[Boolean]])
     case _: CharInfo.type =>
       Code.newInstance[java.lang.Character, Char](x.asInstanceOf[Code[Char]])
+    case _: UnitInfo.type =>
+      Code(x, Code._null[java.lang.Void])
```
Comment (cseed) [inline src/main/scala/is/hail/asm4s/GenericTypeInfo.scala]: Ah, shouldn't Unit/Void take up no stack slots, in which case this is just x.asInstanceOf[Code[Unit]]?

Also, I'm slightly confused how Unit parameters are showing up.

File: src/main/scala/is/hail/asm4s/GenericTypeInfo.scala [hunk 9]
```diff
@@ -34,6 +34,8 @@ final case class GenericTypeInfo[T : TypeInfo]() extends MaybeGenericTypeInfo[T]
         Code.checkcast[java.lang.Boolean](x).invoke[Boolean]("booleanValue").asInstanceOf[Code[T]]
       case _: CharInfo.type =>
         Code.checkcast[java.lang.Character](x).invoke[Char]("charValue").asInstanceOf[Code[T]]
+      case _: UnitInfo.type =>
+        Code.toUnit(Code.checkcast[java.lang.Void](x)).asInstanceOf[Code[T]]
```
Comment (cseed) [inline src/main/scala/is/hail/asm4s/GenericTypeInfo.scala]: Same comment (x produces no values).

File: src/main/scala/is/hail/asm4s/package.scala [hunk 10]
```diff
@@ -119,6 +122,18 @@ package object asm4s {
...
   }
 
+  implicit object UnitInfo extends TypeInfo[Unit] {
+    val name = "V"
+    def loadOp = ???
+    def storeOp = ???
+    def aloadOp = ???
+    def astoreOp = ???
+    val returnOp = RETURN
+    override val slots = 2
```
Comment (cseed) [inline src/main/scala/is/hail/asm4s/package.scala]: Two?  Not 0?

File: src/main/scala/is/hail/asm4s/package.scala [hunk 11]
```diff
@@ -119,6 +122,18 @@ package object asm4s {
...
+
+    def newArray() = new IntInsnNode(NEWARRAY, T_CHAR)
```
Comment (cseed) [inline src/main/scala/is/hail/asm4s/package.scala]: Hmm ... ???  You can't create an array of void in Java.

```
class Foo {
    void main(String[] args) {
        void[] a = new void[10];
    }
}

foo.java:3: error: illegal start of expression
        void[] a = new void[10];
        ^
```

Same code with compiles fine.

File: src/main/scala/is/hail/expr/ir/Compile.scala [hunk 12]
```diff
@@ -31,12 +22,20 @@ object Compile {
 
   type E = Env[(TypeInfo[_], Code[Boolean], Code[_])]
 
-  def apply(ir: IR, fb: FunctionBuilder[_]) {
-    apply(ir, fb, new Env())
+  def toCode(ir: IR, fb: FunctionBuilder[_]): (Code[Unit], Code[Boolean], Code[_]) = {
```
Comment (cseed) [inline src/main/scala/is/hail/expr/ir/Compile.scala]: should this be private[ir], too?

File: src/main/scala/is/hail/expr/ir/ExtractAggregators.scala [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+
+  private def compileAgg2(
+    ir: IR,
+    tAggIn: TAggregable,
+    continuation: (Code[_], Code[Boolean]) => Code[Unit],
+    fb: FunctionBuilder[_],
+    mb: StagedBitSet,
+    env: Compile.E): Code[Unit] = {
+
+    def compileAgg2(ir: IR)(continuation: (Code[_], Code[Boolean]) => Code[Unit]): Code[Unit] =
```
Comment (cseed) [inline src/main/scala/is/hail/expr/ir/ExtractAggregators.scala]: Do you want to just break up the parameters in the original definition and obviate this?

File: src/main/scala/is/hail/asm4s/GenericTypeInfo.scala (hunk 8 above)
Comment (cseed) [reply src/main/scala/is/hail/asm4s/GenericTypeInfo.scala]: Thanks for the explanation, this all makes sense now.  I'm on board.

File: src/main/scala/is/hail/expr/ir/ExtractAggregators.scala [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+  private def emitAgg2(
+    ir: IR,
+    tAggIn: TAggregable,
+    continuation: (Code[_], Code[Boolean]) => Code[Unit],
+    fb: FunctionBuilder[_],
+    mb: StagedBitSet,
+    env: Emit.E): Code[Unit] = {
+
+    def emitAgg2(ir: IR)(continuation: (Code[_], Code[Boolean]) => Code[Unit]): Code[Unit] =
+      this.emitAgg2(ir, tAggIn, continuation, fb, mb, env)
```
Comment (cseed) [inline src/main/scala/is/hail/expr/ir/ExtractAggregators.scala]: I just mean, you define a variant of `emitAgg2` with the same parameters but multiple parameter lists.  Why not just make the original definition have multiple parameter lists?  Then you can delete the variant.

File: src/main/scala/is/hail/annotations/aggregators/RegionValueSumAggregator.scala [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+import is.hail.asm4s._
+import is.hail.expr._
+import is.hail.expr.ir._
+import is.hail.annotations._
+import is.hail.expr.RegionValueAggregator
+
+import scala.reflect.ClassTag
+
+object RegionValueSumAggregator {
+  private def seqOp[Agg >: Null : ClassTag : TypeInfo, T : ClassTag](t: Type): (Code[RegionValueAggregator], Code[_], Code[Boolean]) => Code[Unit] =
```
Comment (cseed) [inline src/main/scala/is/hail/annotations/aggregators/RegionValueSumAggregator.scala]: `SeqOp[Agg >: RegionValueAggregator, ...` is more specific?

File: src/main/scala/is/hail/expr/ir/ExtractAggregators.scala (hunk 14 above)
Comment (danking) [reply src/main/scala/is/hail/expr/ir/ExtractAggregators.scala]: This one has fewer parameters, just `ir` and `continuation`. The other parameters never change, so the nested function allows me to avoid repeating `(..., fb, mb, env)` everywhere. I elide the invariant arguments.

File: src/main/scala/is/hail/annotations/aggregators/RegionValueSumAggregator.scala (hunk 15 above)
Comment (danking) [reply src/main/scala/is/hail/annotations/aggregators/RegionValueSumAggregator.scala]: The bound would be `Agg >: Null <: RegionValueAggregator`, I'll add this in the `ApplyAggOp` PR.

Comment (cseed) [reply src/main/scala/is/hail/annotations/aggregators/RegionValueSumAggregator.scala]: Wait, what?  Agg extends RegionValueAggregator, right?  And U >: T if U extends T, right?  So Agg >: RegionValueAggregator >: Null, that last of which is implied by the RegionValueAggregator bound?  Or am I being stupid here?

Comment (danking) [reply src/main/scala/is/hail/annotations/aggregators/RegionValueSumAggregator.scala]: AFAIK, `>:` is read as "is a super type of", so `Agg` should be a subtype of, `<:` RegionValueAggregator (cause it `extends` it) and should be a super type of `Null` (which was a hack I used to indicate that something was a Reference type, I have forgotten why `AnyRef` wasn't the right choice ...)

Comment (danking) [reply src/main/scala/is/hail/annotations/aggregators/RegionValueSumAggregator.scala]: Yeah, here's a Scala page about it https://docs.scala-lang.org/tour/lower-type-bounds.html

Comment (danking) [reply src/main/scala/is/hail/annotations/aggregators/RegionValueSumAggregator.scala]: Honestly, I think we should change the bounds on `CodeObject` to be subtype of `AnyRef`, this `Null` thing is kind of silly.

I think the issue was that `T <: AnyRef` isn't sufficient to write `val t: T = null`, but that's not necessary to call a method.

Comment (cseed) [reply src/main/scala/is/hail/annotations/aggregators/RegionValueSumAggregator.scala]: It builds just fine for me with just `Agg <: RegionValueAggregator`.

You need it to be a subtype of `AnyRef` for the checkcast to work and you get that for free from `RegionValueAggregator`.  I do admit I'm not sure how the `Null` constraint on `CodeObject` is getting satisfied in this case.

I still advocate adding a region value aggregator bound.

Comment (danking) [reply src/main/scala/is/hail/annotations/aggregators/RegionValueSumAggregator.scala]: Yeah RegionValueAggregator bound makes the most sense. I'll get rid of Null in the AggOp PR.

**Comment by cseed** [comment]
Still needs a rebase.

**Comment by danking** [comment]
Master changes faster than I can rebase

**Comment by danking** [comment]
rebased.
